### PR TITLE
Remediate TunnelVision, TunnelCrack and fix "Exclude Local Networks"

### PR DESCRIPTION
**Pull request by diegoreymendez**
Task/Issue URL: https://app.asana.com/0/1206580121312550/1208686409805161/f
Tech Design URL: https://app.asana.com/0/481882893211075/1208643192597095/f

iOS PR: https://github.com/duckduckgo/iOS/pull/3460
macOS PR: https://github.com/duckduckgo/macos-browser/pull/3422
What kind of version bump will this require?: Major

## Description

Remediate TunnelVision, TunnelCrack and fix "Exclude Local Networks".

## Testing

Please refer to the platform-specific PRs for testing instructions.

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: Sources/BrowserServicesKit/InternalUserDecider/UserDefaults+isInternalUser.swift
Comment (diegoreymendez) [inline Sources/BrowserServicesKit/InternalUserDecider/UserDefaults+isInternalUser.swift:22]: `UserDefaults` can now be used to store the internal user state so that agents and extensions can access said state.

File: Sources/NetworkProtection/Models/AnyIPAddress.swift
Comment (diegoreymendez) [inline Sources/NetworkProtection/Models/AnyIPAddress.swift:22]: This was causing issues because some IP addresses contained by this were converting to empty strings.  I'm removing the `IPAddress` protocol and consumers will instead need to call `ipAddress` on an instance of this class.

File: Sources/NetworkProtection/NetworkProtectionDeviceManager.swift
Comment (diegoreymendez) [inline Sources/NetworkProtection/NetworkProtectionDeviceManager.swift:69]: Included and excluded routes will no longer by passed by the client or by settings.  They'll be calculated by the extension based on settings.

If we ever need to add support for custom IP ranges, we'll just add the appropriate overrides in settings but still calculate all the routing here.

File: Sources/NetworkProtection/NetworkProtectionDeviceManager.swift
Comment (diegoreymendez) [inline Sources/NetworkProtection/NetworkProtectionDeviceManager.swift:174]: This bool is not necessary and not really useful.  Enforce routes is sort of our kill switch now.

File: Sources/NetworkProtection/NetworkProtectionDeviceManager.swift
Comment (diegoreymendez) [inline Sources/NetworkProtection/NetworkProtectionDeviceManager.swift:292]: I don't think this can contain PII - it'll mostly be mostly static IP ranges and DNS ip.

File: Sources/NetworkProtection/NetworkProtectionDeviceManager.swift
Comment (diegoreymendez) [inline Sources/NetworkProtection/NetworkProtectionDeviceManager.swift:287]: This class takes care of calculating the appropriate routes.  Makes it really easy to write unit tests as well.

I'll try to get some in before we merge this.

File: Sources/NetworkProtection/PacketTunnelProvider.swift
Comment (diegoreymendez) [inline Sources/NetworkProtection/PacketTunnelProvider.swift:794]: Routes are no longer passed into the `PacketTunnelProvider`.

File: Sources/NetworkProtection/PacketTunnelProvider.swift
Comment (diegoreymendez) [inline Sources/NetworkProtection/PacketTunnelProvider.swift:1139]: Not sure if the extension or main app should decide when to restart the adapter.  Maybe I should just do it here?

But doing it here prevents us from doing batch changes - although to be honest we don't do them :)

File: Sources/NetworkProtection/PacketTunnelProvider.swift
Comment (diegoreymendez) [inline Sources/NetworkProtection/PacketTunnelProvider.swift:1240]: Convenience method for restarting the adapter very quickly and without going through server re-registration.

File: Sources/NetworkProtection/Status/ConnectionErrorObserver/ConnectionErrorObserverThroughSession.swift
Comment (diegoreymendez) [inline Sources/NetworkProtection/Status/ConnectionErrorObserver/ConnectionErrorObserverThroughSession.swift:100]: Oh I should enable this back I think?  But strangely enough this has eliminated all the crashing I was seeing.

Comment (samsymons) [reply Sources/NetworkProtection/Status/ConnectionErrorObserver/ConnectionErrorObserverThroughSession.swift:100]: This is an interesting one because I've been having problems with it on iOS today as well. Let's sync about it on Tuesday after Apple Weekly.

Comment (diegoreymendez) [reply Sources/NetworkProtection/Status/ConnectionErrorObserver/ConnectionErrorObserverThroughSession.swift]: Re-enabled.

File: Sources/NetworkProtection/PacketTunnelProvider.swift
Comment (diegoreymendez) [reply Sources/NetworkProtection/PacketTunnelProvider.swift]: Moved to the macOS target.

File: Sources/NetworkProtection/PacketTunnelProvider.swift
Comment (diegoreymendez) [reply Sources/NetworkProtection/PacketTunnelProvider.swift]: Moved to the macOS target because `NETP_SYSTEM_EXTENSION` isn't defined in BSK.

File: Sources/NetworkProtection/Routing/VPNRoutingRange.swift
Comment (not-a-rootkit) [inline Sources/NetworkProtection/Routing/VPNRoutingRange.swift:38]: Is this commented out because it clashes with our VPN tunnel IP range? Should we add ranges surrounding the tunnel range?

File: Tests/NetworkProtectionTests/Mocks/NetworkProtectionServerMocks.swift
Comment (not-a-rootkit) [inline Tests/NetworkProtectionTests/Mocks/NetworkProtectionServerMocks.swift:21]: Should this be removed?

**Review by not-a-rootkit — APPROVED** [review]
LGTM, just some open questions on the 10.0.0.0 IP range and a nit with left-in commented-out code. 

Not a blocker, but open question: do you think there is a way to test the routing table in isolation? I don't see an obvious way that doesn't involve quite a bit of work, but you know more than me :)

File: Sources/NetworkProtection/Routing/VPNRoutingRange.swift
Comment (diegoreymendez) [reply Sources/NetworkProtection/Routing/VPNRoutingRange.swift:38]: I'm matching Android here, where they exclude the 10.x.x.x range because it's used for DNS and generally reserved for out VPN.

I do think we should included everything but the DNS but since Android is doing this, I'm assuming it's good enough for us, for now.

File: Tests/NetworkProtectionTests/Mocks/NetworkProtectionServerMocks.swift
Comment (diegoreymendez) [reply Tests/NetworkProtectionTests/Mocks/NetworkProtectionServerMocks.swift]: Yes 100%.

**Pull request merged**
